src/aCorr.cu

```diff
@@ -0,0 +1,358 @@
+ *   from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* 
+
+ 
+
+Implements the grid-Correlation onto a GPU using CUDA. 
+
+*/
+#include <iostream>
+#include "bifrost/aCorr.h"
+#include "assert.hpp"
+#include "trace.hpp"
+#include "utils.hpp"
+#include "cuda.hpp"
+#include "cuda/stream.hpp"
+//#include <complex.h>
+#include "Complex.hpp"
+
+#include <thrust/device_vector.h>
+
+#define tile 256 // Number of threads per thread-block
+
+/********************/
+/* CUDA ERROR CHECK */
+/********************/
+#define gpuErrchk(ans) { gpuAssert((ans), __FILE__, __LINE__); }
+inline void gpuAssert(cudaError_t code, const char *file, int line, bool abort=true)
+{
+   if (code != cudaSuccess) 
+   {
+      fprintf(stderr,"GPUassert: %s %s %d\n", cudaGetErrorString(code), file, line);
+      if (abort) exit(code);
+   }
+}
+
+
+
+inline
+cudaError_t checkCuda(cudaError_t result)
+{
+#if defined(DEBUG) || defined(_DEBUG)
+  if (result != cudaSuccess) {
+    fprintf(stderr, "CUDA Runtime Error: %s\n", cudaGetErrorString(result));
+    assert(result == cudaSuccess);
+  }
+#endif
+  return result;
+}
+
+struct __attribute__((aligned(1))) nibble2 {
+    // Yikes!  This is dicey since the packing order is implementation dependent!  
+    signed char y:4, x:4;
+};
+
+struct __attribute__((aligned(1))) blenib2 {
+    // Yikes!  This is dicey since the packing order is implementation dependent!
+    signed char x:4, y:4;
+};
+
+template<typename In, typename Out>
+__global__ void ACorr(int nbaseline, int npol, int nbatch, int nchan,
+		     In* d_in,
+		     Out* d_out){
+
+        int bid_x = blockIdx.x, bid_y = blockIdx.y, bid_z = blockIdx.z ;
+        int blk_x = blockDim.x;
+        int grid_x = gridDim.x, grid_y = gridDim.y, grid_z = gridDim.z ;
+        int tid_x = threadIdx.x ;
+        int pol_skip = grid_z*blk_x;
+       // Making use of shared memory for faster memory accesses by the threads
+        
+	extern  __shared__ float2 shared[] ;
+        In* xx = reinterpret_cast<In *>(shared);
+        In* yy= xx + blk_x; 	
+
+        // Access pattern is such that coaelescence is achieved both for read and writes to global and shared memory
```

Comment:
it seems like the only difference between `ACorr` and `XGrid_kernel` is a slightly different calculation for `bid` and `bid2`. Can these be combined into a single kernel?

Reply:
> it seems like the only difference between `ACorr` and `XGrid_kernel` is a slightly different calculation for `bid` and `bid2`. Can these be combined into a single kernel?

These have to be separate because ACorr's input is essentially the raw X and Y voltages to estimate the autocorrelation for every antenna for each frequency channel, time-stamp and polarization. So the input and output sequence(or array shapes) are different from that of XGrid which takes the X and Y uv-grids as input.